packager/VisibilityChecker.cc

```diff
@@ -0,0 +1,558 @@
+                if (!member.exists() || !member.isClassOrModule()) {
+                    testSym = core::Symbols::noClassOrModule();
+                    break;
+                }
+
+                testSym = member.asClassOrModuleRef();
+            }
+
+            if (testSym.exists()) {
+                auto file = testSym.data(ctx)->loc().file();
+                if (db.getPackageNameForFile(file) != this->package.mangledName()) {
+                    testSym.data(ctx)->addLoc(ctx, ctx.locAt(loc));
+                }
+            }
+        }
+    }
+
+    void checkExportPackage(core::MutableContext &ctx, core::LocOffsets loc, core::NameRef litPackage) {
+        if (litPackage != this->package.mangledName() && !this->package.importsPackage(litPackage)) {
+            if (auto e = ctx.beginError(loc, core::errors::Packager::InvalidExport)) {
+                e.setHeader("Exporting a symbol that isn't owned by this package");
+            }
+        }
+    }
+
+    PropagateVisibility(const core::packages::PackageInfo &package) : package{package} {}
+
+public:
+    // Find uses of export and mark the symbols they mention as exported.
+    ast::ExpressionPtr postTransformSend(core::MutableContext ctx, ast::ExpressionPtr tree) {
+        auto &send = ast::cast_tree_nonnull<ast::Send>(tree);
+        if (send.fun != core::Names::export_()) {
+            return tree;
+        }
+
+        if (send.numPosArgs() != 1) {
+            // an error will have been raised in the packager pass
+            return tree;
+        }
+
+        auto *lit = ast::cast_tree<ast::ConstantLit>(send.getPosArg(0));
+        if (lit == nullptr || lit->symbol == core::Symbols::StubModule()) {
+            // We don't raise an explicit error here, as this is one of two cases:
+            //   1. Export is given a non-constant argument
+            //   2. The argument failed to resolve
+            // In both cases, errors will be raised by previous passes.
+            return tree;
+        }
+
+        auto &db = ctx.state.packageDB();
+        if (lit->symbol.isClassOrModule()) {
+            auto sym = lit->symbol.asClassOrModuleRef();
+            checkExportPackage(ctx, send.loc, db.getPackageNameForFile(sym.data(ctx)->loc().file()));
+            recursiveExportSymbol(ctx, true, sym);
+
+            exportParentNamespace(ctx, sym.data(ctx)->owner);
+        } else if (lit->symbol.isFieldOrStaticField()) {
+            auto sym = lit->symbol.asFieldRef();
+            checkExportPackage(ctx, send.loc, db.getPackageNameForFile(sym.data(ctx)->loc().file()));
+            sym.data(ctx)->flags.isExported = true;
+
+            // When exporting a field, we also export its parent namespace. This is a bit of a hack, and it would be
+            // great to remove this, but this was the behavior of the previous packager implementation.
+            // TODO: can we get away with only marking the owner as exported, and not everything up to the package?
```

Comment:
Another TODO. Should we add this to the followup doc?

Reply:
I'll take another look and add it if it's still relevant.

Reply:
I'll add an entry to the followup doc, I think this is worth investigating but there's a lot of reliance on it in the Stripe codebase right now.